keps/sig-network/2079-network-policy-port-range/README.md

```diff
@@ -0,0 +1,306 @@
+    - protocol: TCP
+      port: 32001
+    - protocol: TCP
+      port: 32002
+    - protocol: TCP
+      port: 32003
+[...]
+    - protocol: TCP
+      port: 32768
+```
+
+So for the user:
+* To allow a range of ports, each of them must be declared as an item from 
+``ports`` array
+* To make an exception needs a declaration of all ports but the exception
+
+Adding a new ``PortRange`` field inside the ``ports`` will allow a simpler 
+creation of NetworkPolicy to the user.
+
+### Goals
+
+Add Range field to Ports in NetworkPolicy
+
+### Non-Goals
+
+N/A
+
+## Proposal
+
+In NetworkPolicy specification, inside ``NetworkPolicyPort`` object struct 
+specify a new ``Range`` field composed of the minimum and maximum ports 
+inside the range.
+
+If both ``Port`` and ``Range`` are specified, they are cumulative and no further
+validation might occur. It's up to the CNI to summarize both of the fields in a 
+single rule.
+
+### User Stories (Optional)
+
+#### Story 1
+
+I have an application that communicates with NodePorts of a different cluster 
+and I want to allow the egress of the traffic only the NodePort range 
+(eg. 30000-32767) as I don't know which port is going to be allocated on the 
+other side, but don't want to create a rule for each of them.
+
+### Notes/Constraints/Caveats
+
+*  The technology used by the CNI provider might not support port range in a 
+trivial way as described in [#drawbacks]
+
+### Risks and Mitigations
+
+CNIs will need to support the new field in their controllers. For this case 
+we'll try to make broader communication with the main CNIs so they can be aware
+of the new field.
+
+## Design Details
+
+API changes to NetworkPolicy:
+* Add a new struct called ``NetworkPolicyPortRange`` as the following:
+```
+type NetworkPolicyPortRange struct {
+    From         uint16
+    To           uint16
+    Exceptions:  []uint16
+}
+```
+* Add a new field ``spec.ingress|egress.ports.range`` that points to the
+new struct:
+```
+// NetworkPolicyPort describes a port to allow traffic on
+type NetworkPolicyPort struct {
+    Range NetworkPolicyPortRange
+}
+```
+
+### Validations
+The range will need to be validated, with the following scenarios:
+* If there's a ``From`` or a ``To`` field defined, the other one must be defined.
+* ``From`` needs to be less than or equal to ``To``
+* All the ports in the ``Exceptions`` array must be inside the defined range.
+* ``Exception`` can only be defined if ``From`` and ``To`` are also defined.
+* Because ``ports`` is a superset of all ports specified in ``port`` and 
+``range``, if a port is specified in at least one of the fields it should be 
+allowed.
+
+### Test Plan
+
+TBD
+
+### Graduation Criteria
+
+#### Alpha 
+- Add a feature gated new field to NetworkPolicy
+- Communicate CNI providers about the new field
+- Add validation tests in API
+
+#### Beta
+- ``PortRanges`` has been supported for at least 1 minor release
+- At least one CNI provider implements the new field
```

Comment:
Personally I think only requiring one implementation is too loose of a requirement for this field to be on by default. Not sure what the magic number is but should at least be a handful of CNIs?

Reply:
Yes, it was thinking more about having a PR to be reviewed other then a magic number.

I'm tempted to follow the same number proposed by Jay in #2113  and change to:

```
- The name selector has been supported for at least 1 minor release.
- Four commonly used NetworkPolicy (or CNI providers) implement the new field, 
with generally positive feedback on its usage.
- Feature Gate is enabled by Default.```